modules/rostests/kmtests/ntos_io/IoDeviceInterface.c

```diff
@@ -1,125 +1,214 @@
+
+    for (size_t n = 0; n < RTL_NUMBER_OF(Types); ++n)
+    {
+        HANDLE DeviceInterfaceKey, DeviceInterfaceSubKey;
+        OBJECT_ATTRIBUTES ObjectAttributes;
+
+        NTSTATUS Status = IoOpenDeviceInterfaceRegistryKey(&SymbolicLinkName, KEY_CREATE_SUB_KEY, &DeviceInterfaceKey);
+
+        if (NT_SUCCESS(Status))
+        {
+            trace("IoOpenDeviceInterfaceRegistryKey(): success: %d %p\n", n, DeviceInterfaceKey);
+        }
+        else
+        {
+            trace("IoOpenDeviceInterfaceRegistryKey(): fail: %d 0x%x\n", n, Status);
+        }
+
+        InitializeObjectAttributes(&ObjectAttributes,
+                                   &KeyName,
+                                   OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
+                                   DeviceInterfaceKey,
+                                   NULL);
+        Status = ZwCreateKey(&DeviceInterfaceSubKey,
+                             KEY_WRITE,
+                             &ObjectAttributes,
+                             0,
+                             NULL,
+                             REG_OPTION_NON_VOLATILE,
+                             NULL);
+
+        if (NT_SUCCESS(Status))
+        {
+            trace("ZwCreateKey(): successfully created subkey: %d %p\n", n, DeviceInterfaceSubKey);
+
+            ZwDeleteKey(DeviceInterfaceSubKey);
+            ZwClose(DeviceInterfaceSubKey);
+        }
+        else
+        {
+            trace("ZwCreateKey(): failed to create a subkey: %d 0x%x\n", n, Status);
+        }
+
+        ZwClose(DeviceInterfaceKey);
+    }
 }
 
 static
 VOID
-Test_IoRegisterDeviceInterface(VOID)
+Test_IoGetDeviceInterfaceAlias(PWSTR SymbolicLink)
 {
-    GUID Guid = {0x378de44c, 0x56ef, 0x11d1, {0xbc, 0x8c, 0x00, 0xa0, 0xc9, 0x14, 0x05, 0xdd}};
-    DEVICE_OBJECT DeviceObject;
-    EXTENDED_DEVOBJ_EXTENSION DeviceObjectExtension;
-    DEVICE_NODE DeviceNode;
     UNICODE_STRING SymbolicLinkName;
-    NTSTATUS Status;
 
-    RtlInitUnicodeString(&SymbolicLinkName, L"");
+    RtlInitUnicodeString(&SymbolicLinkName, SymbolicLink);
+
+    for (size_t n = 0; n < RTL_NUMBER_OF(Types); ++n)
+    {
+        UNICODE_STRING AliasSymbolicLinkName;
+        NTSTATUS Status = IoGetDeviceInterfaceAlias(&SymbolicLinkName, Types[n], &AliasSymbolicLinkName);
+
+        if (NT_SUCCESS(Status))
+        {
+            trace("IoGetDeviceInterfaceAlias(): success: %d %wZ\n", n, &AliasSymbolicLinkName);
+
+            RtlFreeUnicodeString(&AliasSymbolicLinkName);
+        }
+        else
+        {
+            trace("IoGetDeviceInterfaceAlias(): fail: %d 0x%x\n", n, Status);
+        }
+    }
+}
+
+static
+VOID
+Test_IoSetDeviceInterfaceState(PWSTR SymbolicLink)
+{
+    UNICODE_STRING SymbolicLinkName;
+
+    RtlInitUnicodeString(&SymbolicLinkName, SymbolicLink);
+
+    for (size_t n = 0; n < RTL_NUMBER_OF(Types); ++n)
+    {
+        NTSTATUS Status = IoSetDeviceInterfaceState(&SymbolicLinkName, TRUE);
```

Comment:
You can't just enable a device interface for a device that you don't own. This test has to be done with a device created by the test.

Reply:
So how should I create the device? Maybe you mean device interface? If it's possible only via `IoRegisterDeviceInterface` testcase, then that needs to be implemented too. But I doubt how to implement it properly. In case it's not correct to use this testcase without `IoRegisterDeviceInterface`, then should I disable it for now? :thinking:

Reply:
Ah, just noticed your comment below that it can be done via `IoCreateDevice`. So maybe I can implement that, but I'm not sure. :slightly_frowning_face: